### PR TITLE
[SCRIPTS] Replace finetune_lm.py and sentiment_analysis_cnn.py using Estimator

### PR DESCRIPTION
**Pull request by gongel**
## Description ##
convert https://github.com/dmlc/gluon-nlp/blob/master/scripts/sentiment_analysis/finetune_lm.py by using Estimator.

### Essentials ###
- [x] PR's title starts with a category (e.g. [BUGFIX], [MODEL], [TUTORIAL], [FEATURE], [DOC], etc)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

cc @dmlc/gluon-nlp-team

### CODE REVIEW

**Comment by mli** [comment]
Job PR-1070/1 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-1070/1/index.html

**Comment by leezu** [comment]
Thanks @gongel. Please replace the exiting file with the updated file, instead of introducing a new one. Please ask @liuzh91 or @zburning if you're unsure how to do it with git.

**Comment by liuzh91** [comment]
1. You should move MyBatchProcessor to gluonnlp api with a more appropriate name
2. You can also create a class SentimentAnalysisEstimator class in api.

**Comment by gongel** [comment]
All, I got it.

**Comment by mli** [comment]
Job PR-1070/2 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-1070/2/index.html

**Comment by mli** [comment]
Job PR-1070/3 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-1070/3/index.html

File: scripts/sentiment_analysis/finetune_lm.py [hunk 1]
```diff
@@ -94,15 +130,17 @@
     context = mx.cpu()
 else:
     print('Use gpu%d' % args.gpu)
-    context = mx.gpu(args.gpu)
+    context = mx.gpu(int(args.gpu))
```
Comment (leezu) [inline scripts/sentiment_analysis/finetune_lm.py]: The int cast is not required because `type=int` is already specified in argparse.

File: scripts/sentiment_analysis/finetune_lm.py [hunk 2]
```diff
@@ -193,152 +242,131 @@ def preprocess_dataset(dataset):
...
+est = Estimator(net=net,
+                loss=loss_fn,
+                train_metrics=[mx.metric.Loss()],
+                val_metrics=[mx.metric.Accuracy()],
+                trainer=trainer,
+                context=context,
+                batch_processor=SentimentAnalysisBatchProcessor())
+
+with warnings.catch_warnings():
+    warnings.simplefilter("ignore")
```
Comment (leezu) [inline scripts/sentiment_analysis/finetune_lm.py]: Why ignore warnings?

File: scripts/sentiment_analysis/finetune_lm.py [hunk 3]
```diff
@@ -193,152 +242,131 @@ def preprocess_dataset(dataset):
...
-
-    net.load_parameters(glob.glob(args.save_prefix+'_*.params')[-1], context)
-    valid_avg_L, valid_acc = evaluate(valid_dataloader)
-    test_avg_L, test_acc = evaluate(test_dataloader)
-    print('Best validation loss %g, validation acc %.4f'%(valid_avg_L, valid_acc))
-    print('Best test loss %g, test acc %.4f'%(test_avg_L, test_acc))
-    print('Total time cost %.2fs'%(time.time()-start_pipeline_time))
-
-
-if __name__ == '__main__':
```
Comment (leezu) [inline scripts/sentiment_analysis/finetune_lm.py]: It's good to keep the if condition here. Instead of the `train()`, you can put the `est.fit()` inside.

**Comment by mli** [comment]
Job PR-1070/4 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-1070/4/index.html

File: scripts/sentiment_analysis/finetune_lm.py [hunk 4]
```diff
@@ -193,152 +241,130 @@ def preprocess_dataset(dataset):
...
+                for data, valid_length in zip(data_list, valid_length_list)
+            ]
+            loss = [
+                estimator.loss(y_hat, y).mean()
+                for y_hat, y in zip(pred, label_list)
+            ]
+        for l in loss:
+            l.backward()
+        # Clip gradient
+        if args.clip is not None:
```
Comment (liuzh91) [inline scripts/sentiment_analysis/finetune_lm.py]: You should process gradient clipping and updating in the `GradientUpdateHandler`.  In current estimator framework, `fit_batch` only computes the forwarding process, i.e., pass the input through the network to get the predictions and then calculates the loss. All the gradient backpropagation are done in the `GradientUpdateHandler`. See line 714 for future reference. https://github.com/apache/incubator-mxnet/blob/master/python/mxnet/gluon/contrib/estimator/event_handler.py#L714

File: scripts/sentiment_analysis/finetune_lm.py [hunk 5]
```diff
@@ -193,152 +241,130 @@ def preprocess_dataset(dataset):
...
+            grads = [
+                p.grad(estimator.context)
+                for p in self.net.collect_params().values()
+            ]
+            gluon.utils.clip_global_norm(grads, args.clip)
```
Comment (liuzh91) [inline scripts/sentiment_analysis/finetune_lm.py]: Because you split data and labels across multiple contexts, you probably need to call `allreduce_grads()` after gradient clipping before back propogation.

**Comment by eric-haibin-lin** [comment]
@gongel  Please complete the checklist 

> Essentials
>  PR's title starts with a category (e.g. [BUGFIX], [MODEL], [TUTORIAL], [FEATURE], [DOC], etc)
>  Changes are complete (i.e. I finished coding on this PR)
>  All changes have test coverage
>  Code is well-documented
>

**Comment by gongel** [comment]
> @gongel Please complete the checklist
> 
> > Essentials
> > PR's title starts with a category (e.g. [BUGFIX], [MODEL], [TUTORIAL], [FEATURE], [DOC], etc)
> > Changes are complete (i.e. I finished coding on this PR)
> > All changes have test coverage
> > Code is well-documented

Ok, finished

File: scripts/sentiment_analysis/sentiment_analysis_cnn.py [hunk 6]
```diff
@@ -71,138 +95,166 @@
...
+                # metric wrapper for loss values
+                metric.update(0, loss)
+            else:
+                metric.update(label, pred)
+
+    def epoch_end(self, estimator, *args, **kwargs):
+        for metric in self.metrics:
+            metric_name, metric_val = metric.get()
+            if 'validation' in metric_name:
+                self.metric_history.setdefault(metric_name,
```
Comment (liuzh91) [inline scripts/sentiment_analysis/sentiment_analysis_cnn.py]: I understand the reason your create this metric handler is for tracking the history of metrics. I think you may create a standalone EventHandler for the metric logging, e.g., MetricHistoryHandler(EpochEnd). MetricHandler is used to compute the metric of current batch while MetricHistoryHandler is used to tracked history.

File: scripts/sentiment_analysis/sentiment_analysis_cnn.py [hunk 7]
```diff
@@ -71,138 +95,166 @@
...
+                                               []).append(metric_val)
+
+
+class SentimentAnalysisCNNBatchProcessor(BatchProcessor):
```
Comment (liuzh91) [inline scripts/sentiment_analysis/sentiment_analysis_cnn.py]: In the future, the batch processor should be moved into estimator api. It is OK to keep the class here temporarily.

File: scripts/sentiment_analysis/sentiment_analysis_cnn.py [hunk 8]
```diff
@@ -71,138 +95,166 @@
...
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
```
Comment (liuzh91) [inline scripts/sentiment_analysis/sentiment_analysis_cnn.py]: I believe the default batch processor does not have any arguments and it is constructor is just a pass, why do you call the super class constructor here?

File: scripts/sentiment_analysis/sentiment_analysis_cnn.py [hunk 9]
```diff
@@ -71,138 +95,166 @@
...
+                trainer=trainer,
+                context=context,
+                evaluation_loss=loss_fn,
+                val_metrics=[mx.metric.Accuracy()],
+                batch_processor=SentimentAnalysisCNNBatchProcessor())
 if __name__ == '__main__':
     if args.data_name == 'TREC':
-        train(model, train_dataset, test_dataset)
+        random.shuffle(train_dataset)
+        sp = len(train_dataset) // 10
```
Comment (liuzh91) [inline scripts/sentiment_analysis/sentiment_analysis_cnn.py]: what is this 10 means? Better give it a proper name here.

File: scripts/sentiment_analysis/sentiment_analysis_cnn.py (hunk 6 above)
Comment (gongel) [reply scripts/sentiment_analysis/sentiment_analysis_cnn.py]: Yeah, I want to track the history of metrics at EpochEnd. I will fix it in the next commit.

File: scripts/sentiment_analysis/sentiment_analysis_cnn.py [hunk 10]
```diff
@@ -3,11 +3,12 @@
 ===============================================
 This example shows how to use convolutional neural networks (textCNN)
 for sentiment analysis on various datasets.
-
 Kim, Y. (2014). Convolutional neural networks for sentence classification.
 arXiv preprint arXiv:1408.5882.
 """
 
+# coding: utf-8
```
Comment (leezu) [inline scripts/sentiment_analysis/sentiment_analysis_cnn.py]: This is only needed in Python 2. But we don't support Python 2, so you can remove it.

Comment (gongel) [reply scripts/sentiment_analysis/sentiment_analysis_cnn.py]: Ok, done.

**Comment by leezu** [comment]
@gongel I merged master branch into your development branch, as it contains fixes to allow the CI to run properly.

**Comment by mli** [comment]
Job PR-1070/10 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-1070/10/index.html

**Comment by leezu** [comment]
@gongel please fix the linting issues

```[2020-01-20T23:20:46.000Z] ************* Module scripts.sentiment_analysis.sentiment_analysis_cnn

[2020-01-20T23:20:46.000Z] scripts/sentiment_analysis/sentiment_analysis_cnn.py:96:0: C0301: Line too long (110/100) (line-too-long)

[2020-01-20T23:20:46.000Z] scripts/sentiment_analysis/sentiment_analysis_cnn.py:98:0: C0301: Line too long (128/100) (line-too-long)

[2020-01-20T23:20:46.000Z] scripts/sentiment_analysis/sentiment_analysis_cnn.py:101:0: C0301: Line too long (159/100) (line-too-long)

[2020-01-20T23:20:46.000Z] scripts/sentiment_analysis/sentiment_analysis_cnn.py:120:16: C4001: Invalid string quote ", should be ' (invalid-string-quote)

[2020-01-20T23:20:46.000Z] scripts/sentiment_analysis/sentiment_analysis_cnn.py:121:16: C4001: Invalid string quote ", should be ' (invalid-string-quote)

[2020-01-20T23:20:46.000Z] scripts/sentiment_analysis/sentiment_analysis_cnn.py:108:0: C0116: Missing function or method docstring (missing-function-docstring)

[2020-01-20T23:20:46.000Z] scripts/sentiment_analysis/sentiment_analysis_cnn.py:125:0: C0115: Missing class docstring (missing-class-docstring)

[2020-01-20T23:20:46.000Z] scripts/sentiment_analysis/sentiment_analysis_cnn.py:131:0: W0621: Redefining name 'args' from outer scope (line 85) (redefined-outer-name)

[2020-01-20T23:20:46.000Z] scripts/sentiment_analysis/sentiment_analysis_cnn.py:139:0: C0115: Missing class docstring (missing-class-docstring)

[2020-01-20T23:20:46.000Z] scripts/sentiment_analysis/sentiment_analysis_cnn.py:140:4: W0231: __init__ method from base class 'BatchProcessor' is not called (super-init-not-called)

[2020-01-20T23:20:46.000Z] scripts/sentiment_analysis/sentiment_analysis_cnn.py:183:8: W0621: Redefining name 'train_dataloader' from outer scope (line 224) (redefined-outer-name)

[2020-01-20T23:20:46.000Z] scripts/sentiment_analysis/sentiment_analysis_cnn.py:186:8: W0621: Redefining name 'val_dataloader' from outer scope (line 227) (redefined-outer-name)

[2020-01-20T23:20:46.000Z] scripts/sentiment_analysis/sentiment_analysis_cnn.py:181:8: W0621: Redefining name 'train_dataset' from outer scope (line 96) (redefined-outer-name)

[2020-01-20T23:20:46.000Z] scripts/sentiment_analysis/sentiment_analysis_cnn.py:189:8: W0621: Redefining name 'est' from outer scope (line 212) (redefined-outer-name)

[2020-01-20T23:20:46.000Z] scripts/sentiment_analysis/sentiment_analysis_cnn.py:173:0: C0116: Missing function or method docstring (missing-function-docstring)

[2020-01-20T23:20:46.000Z] scripts/sentiment_analysis/sentiment_analysis_cnn.py:36:0: W0611: Unused EpochBegin imported from mxnet.gluon.contrib.estimator.event_handler (unused-import)

[2020-01-20T23:20:46.000Z] scripts/sentiment_analysis/sentiment_analysis_cnn.py:36:0: W0611: Unused BatchEnd imported from mxnet.gluon.contrib.estimator.event_handler (unused-import)

[2020-01-20T23:20:46.000Z] ************* Module scripts.sentiment_analysis.finetune_lm

[2020-01-20T23:20:46.000Z] scripts/sentiment_analysis/finetune_lm.py:37:0: C0301: Line too long (118/100) (line-too-long)

[2020-01-20T23:20:46.000Z] scripts/sentiment_analysis/finetune_lm.py:288:0: C0115: Missing class docstring (missing-class-docstring)

[2020-01-20T23:20:46.000Z] scripts/sentiment_analysis/finetune_lm.py:289:4: W0231: __init__ method from base class 'BatchProcessor' is not called (super-init-not-called)

[2020-01-20T23:20:46.000Z] scripts/sentiment_analysis/finetune_lm.py:335:0: C0115: Missing class docstring (missing-class-docstring)

[2020-01-20T23:20:46.000Z] scripts/sentiment_analysis/finetune_lm.py:340:0: W0621: Redefining name 'args' from outer scope (line 123) (redefined-outer-name)
```

**Comment by leezu** [comment]
Ping @gongel

**Comment by gongel** [comment]
> Ping @gongel

Done.

File: scripts/sentiment_analysis/finetune_lm.py [hunk 11]
```diff
@@ -333,11 +335,12 @@ def fit_batch(self, estimator, train_batch, batch_axis=0):
 
 
 class ClipGradientHandler(GradientUpdateHandler):
+    """handler for clipping gradient"""
     def __init__(self, clip=None, **kwargs):
         super().__init__(**kwargs)
         self.clip = clip
 
-    def batch_end(self, estimator, *args, **kwargs):
+    def batch_end(self, estimator, *argss, **kwargs):
```
Comment (leezu) [inline scripts/sentiment_analysis/finetune_lm.py]: typo?

File: scripts/sentiment_analysis/finetune_lm.py [hunk 12]
```diff
@@ -286,8 +287,9 @@ def preprocess_dataset(dataset):
 
 
 class FituneLmBatchProcessor(BatchProcessor):
+    """subclass for FituneLm"""
     def __init__(self):
```
Comment (leezu) [inline scripts/sentiment_analysis/finetune_lm.py]: You can delete this function. It's currently equivalent to being deleted.

**Comment by leezu** [comment]
@gongel Thanks. Please see above comments.

File: scripts/sentiment_analysis/finetune_lm.py (hunk 11 above)
Comment (gongel) [reply scripts/sentiment_analysis/finetune_lm.py]: Due to redefining name 'args' from outer scope (line 123) (redefined-outer-name), I renamed args to argss

Comment (leezu) [reply scripts/sentiment_analysis/finetune_lm.py]: I see. Thank you for clarifying. One alternative is to move the problematic outer scope statements to the end of the file, inside the if statement.
In general it's bad practice to do any "actual computation" outside of the if statement at the end of the file.

Comment (gongel) [reply scripts/sentiment_analysis/finetune_lm.py]: OK, I get it. By the way, can I disable the linting issue(W0621)?

**Comment by mli** [comment]
Job PR-1070/12 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-1070/12/index.html

**Comment by eric-haibin-lin** [comment]
what is the status of this?

**Comment by mli** [comment]
Job PR-1070/14 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-1070/14/index.html

File: scripts/sentiment_analysis/finetune_lm.py [hunk 13]
```diff
@@ -23,19 +22,21 @@
...
 from mxnet.gluon import HybridBlock
 from mxnet.gluon.data import DataLoader
-
+from mxnet.gluon.contrib.estimator import Estimator
+from mxnet.gluon.contrib.estimator.event_handler import EarlyStoppingHandler, \
+	CheckpointHandler, GradientUpdateHandler
+from mxnet.gluon.contrib.estimator.batch_processor import BatchProcessor
 import gluonnlp as nlp
 
 nlp.utils.check_version('0.7.0')
```
Comment (leezu) [inline scripts/sentiment_analysis/finetune_lm.py]: Should be 0.9

**Comment by mli** [comment]
Job PR-1070/15 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-1070/15/index.html

**Comment by StrayBird-ATSH** [comment]
@eric-haibin-lin @sxjscience  Should this be updated in any aspect? Anything should be fixed to merge this pull request?
